File: exercises/C#/03 PhotoSearcher/03 PhotoSearcher/FuzzyDate.cs
using System;

namespace _03_PhotoSearcher
{
    public class FuzzyDate
    {
        private readonly DateTime startDate;
        private readonly DateTime endDate;

        public enum Quarters
        {
            Q1,
            Q2,
            Q3,
            Q4
        }

        public enum YearType
        {
            Start,
            Halfway,
            End
        }

        public FuzzyDate()
        {
            startDate = new DateTime(0);
            endDate = new DateTime(0);
        }

        public FuzzyDate(string str)
        {
            if (str == "now")
            {
                startDate = DateTime.Now;
                endDate = DateTime.Now;
            }
        }

        public FuzzyDate(Quarters quart, int year)
        {
            switch (quart)
            {
                case Quarters.Q1:
                    startDate = new DateTime(year, 1, 1);
                    endDate = new DateTime(year, 3, 31);
                    break;
                case Quarters.Q2:
                    startDate = new DateTime(year, 4, 1);
                    endDate = new DateTime(year, 6, 30);
                    break;
                case Quarters.Q3:
                    startDate = new DateTime(year, 7, 1);
                    endDate = new DateTime(year, 9, 30);
                    break;
                case Quarters.Q4:
                    startDate = new DateTime(year, 10, 1);
                    endDate = new DateTime(year, 12, 31);
                    break;
            }
        }

        public FuzzyDate(YearType type, int year)
        {
            switch (type)
            {
                case YearType.Start:
                    startDate = new DateTime(year, 1, 1);
                    endDate = new DateTime(year, 4, 30);
                    break;
                case YearType.Halfway:
                    startDate = new DateTime(year, 5, 1);
                    endDate = new DateTime(year, 8, 31);
                    break;
                case YearType.End:
                    startDate = new DateTime(year, 9, 1);
                    endDate = new DateTime(year, 12, 31);
                    break;
            }
        }

        public DateTime GetFuzzyStart()
        {
            return startDate;
        }

        public DateTime GetFuzzyEnd()
        {
            return endDate;
        }

        public DateTime GetFuzzyRange()
        {
            return new DateTime(endDate.Ticks - startDate.Ticks);
        }

        public bool Contains(DateTime lookUpDate)
        {
            var startContains = startDate.CompareTo(lookUpDate) <= 0;
            var endContains = endDate.CompareTo(lookUpDate) >= 0;
            return (startContains && endContains);
        }
    }
}

File: exercises/03 PhotoSearcher/C++/03 PhotoSearcher/FuzzyDate.cpp
#include "FuzzyDate.h"

using namespace std;
using namespace date;

FuzzyDate::FuzzyDate() :
	_startDate(year{1970} / 1 / 1),
	_endDate(year{ 1970 } / 1 / 1)
{
}

FuzzyDate::FuzzyDate(const string& str) :
	FuzzyDate()
{
	if (str == "now")
	{
		auto today = year_month_day{ floor<days>(chrono::system_clock::now()) };
		_startDate = today;
		_endDate = today;
	}
}

FuzzyDate::FuzzyDate(Quarters quart, int year) :
	FuzzyDate()
{
	switch (quart)
	{
	case Quarters::Q1:
		_startDate = ::year{ year } / 1 / 1;
		_endDate = ::year{ year } / 3 / 31;
		break;
	case Quarters::Q2:
		_startDate = ::year{ year } / 4 / 1;
		_endDate = ::year{ year } / 6 / 30;
		break;
	case Quarters::Q3:
		_startDate = ::year{ year } / 7 / 1;
		_endDate = ::year{ year } / 9 / 30;
		break;
	case Quarters::Q4:
		_startDate = ::year{ year } / 10 / 1;
		_endDate = ::year{ year } / 12 / 31;
		break;
	}
}

FuzzyDate::FuzzyDate(YearType type, int year) :
	FuzzyDate()
{
	switch (type)
	{
	case YearType::START:
		_startDate = ::year{ year } / 1 / 1;
		_endDate = ::year{ year }/ 4 / 30;
		break;
	case YearType::HALFWAY:
		_startDate = ::year{ year } / 5 / 1;
		_endDate = ::year{ year } / 8 / 31;
		break;
	case YearType::END:
		_startDate = ::year{ year } / 9 / 1;
		_endDate = ::year{ year } / 12 / 31;
		break;
	}
}


FuzzyDate::~FuzzyDate()
{
}

const date::year_month_day & FuzzyDate::getFuzzyStart() const
{
	return _startDate;
}

const date::year_month_day & FuzzyDate::getFuzzyEnd() const
{
	return _endDate;
}

date::days FuzzyDate::getFuzzyRange() const
{
	return sys_days(_endDate) - sys_days(_startDate);
}

bool FuzzyDate::contains(const date::year_month_day& lookupDate) const
{
	return lookupDate >= _startDate && lookupDate <= _endDate;
}
